File: python/array/longest_common_prefix.py
"""
Problem 7: Longest Common Prefix

Write a function to find the longest common prefix string amongst an array of strings.
If there is no common prefix, return an empty string.

Approaches:
1. Vertical scan — shrink a running prefix until every string starts with it: O(n·m) time, O(1) space
2. Sort — sort lexicographically, then compare the first and last strings: O(n·m log n) time, O(n·m) space
"""


def longest_common_prefix_naive(strings: list[str]) -> str:
    # Start with strings[0] as prefix; shrink from the end until all strings match.
    # Time: O(n·m)  Space: O(1)
    if not strings:
        return ""
    strings_length: int = len(strings)
    if strings_length == 1:
        return strings[0]

    common_prefix: str = strings[0]
    string: str
    for string in strings[1:]:
        while not string.startswith(common_prefix):
            common_prefix = common_prefix[:-1]
            if not common_prefix:
                return ""
    return common_prefix


def longest_common_optimal(strings: list[str]) -> str:
    # Sort lexicographically; LCP of all strings equals LCP of sorted endpoints.
    # Time: O(n·m log n)  Space: O(n·m)
    if not strings:
        return ""
    strings_length: int = len(strings)
    if strings_length == 1:
        return strings[0]

    sorted_strings: list[str] = sorted(strings)
    first: str = sorted_strings[0]
    last: str = sorted_strings[-1]
    first_length: int = len(first)
    last_length: int = len(last)

    common_prefix: str = ""
    idx: int
    for idx in range(first_length):
        if idx < last_length and first[idx] == last[idx]:
            common_prefix += first[idx]
        else:
            break
    return common_prefix


def main() -> None:
    strings: list[str] = list(input().split(" "))

    common_prefix: str = longest_common_prefix_naive(strings)
    print(common_prefix)

    common_prefix = longest_common_optimal(strings)
    print(common_prefix)


if __name__ == "__main__":
    main()

File: rust/src/bin/array/longest_common_prefix.rs
/*
Test Cases
flower flow flight
dog racecar car
a
bac abc bca
aaa aa aaa

Output (naive / optimal per input line)
fl
fl
"" (empty line)
a
a
"" (empty line)
*/

use std::io::{self, BufRead};

// Start with strings[0] as prefix; shrink from the end until all strings match.
// Time: O(n·m)  Space: O(1)
fn longest_common_prefix_naive(strings: &[String]) -> String {
    if strings.is_empty() {
        return String::new();
    }
    if strings.len() == 1 {
        return strings[0].clone();
    }

    let mut common_prefix = strings[0].clone();
    for string in strings.iter().skip(1) {
        while !string.starts_with(&common_prefix) {
            common_prefix.pop();
            if common_prefix.is_empty() {
                return String::new();
            }
        }
    }
    common_prefix
}

// Sort lexicographically; LCP of all strings equals LCP of sorted endpoints.
// Time: O(n·m log n)  Space: O(n·m)
fn longest_common_optimal(strings: &[String]) -> String {
    if strings.is_empty() {
        return String::new();
    }
    if strings.len() == 1 {
        return strings[0].clone();
    }

    let mut sorted_strings: Vec<String> = strings.to_vec();
    sorted_strings.sort();

    let first = sorted_strings[0].as_str();
    let last = sorted_strings.last().unwrap().as_str();
    let first_chars: Vec<char> = first.chars().collect();
    let last_chars: Vec<char> = last.chars().collect();
    let mut common_prefix = String::new();

    for idx in 0..first_chars.len() {
        if idx < last_chars.len() && first_chars[idx] == last_chars[idx] {
            common_prefix.push(first_chars[idx]);
        } else {
            break;
        }
    }
    common_prefix
}

fn main() {
    let stdin = io::stdin();
    let line = stdin.lock().lines().next().unwrap().unwrap();
    let strings: Vec<String> = line.split_whitespace().map(String::from).collect();

    println!("{}", longest_common_prefix_naive(&strings));
    println!("{}", longest_common_optimal(&strings));
}

#[cfg(test)]
mod tests {
    use super::*;

    fn strs(values: &[&str]) -> Vec<String> {
        values.iter().map(|s| s.to_string()).collect()
    }

    #[test]
    fn test_naive() {
        assert_eq!(
            longest_common_prefix_naive(&strs(&["flower", "flow", "flight"])),
            "fl"
        );
        assert_eq!(
            longest_common_prefix_naive(&strs(&["dog", "racecar", "car"])),
            ""
        );
        assert_eq!(longest_common_prefix_naive(&strs(&["a"])), "a");
        assert_eq!(
            longest_common_prefix_naive(&strs(&["aaa", "aa", "aaa"])),
            "aa"
        );
    }

    #[test]
    fn test_optimal() {
        assert_eq!(
            longest_common_optimal(&strs(&["flower", "flow", "flight"])),
            "fl"
        );
        assert_eq!(
            longest_common_optimal(&strs(&["dog", "racecar", "car"])),
            ""
        );
        assert_eq!(longest_common_optimal(&strs(&["a"])), "a");
        assert_eq!(longest_common_optimal(&strs(&["bac", "abc", "bca"])), "");
        assert_eq!(
            longest_common_optimal(&strs(&["aaa", "aa", "aaa"])),
            "aa"
        );
    }
}
